### PR TITLE
#7107: (10.0.x) fix too-eager closing of multiplexed connections marked as closed but still in use

### LINKED ISSUE #7107: Client timeout and async close exceptions when setting max duration on pool
**Jetty version(s)**

11.0.6

**Java version/vendor** `(use: java -version)`

openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment (build 11.0.11+9-Ubuntu-0ubuntu2.18.04)
OpenJDK 64-Bit Server VM (build 11.0.11+9-Ubuntu-0ubuntu2.18.04, mixed mode, sharing)

**OS type/version**

Linux 5.4.0-89-generic #100~18.04.1-Ubuntu SMP Wed Sep 29 10:59:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Description**

We have been using Jetty in production for a while, client and server side, with a custom client `RoundRobinConnectionPool`  that support TTL connection, like this proposal (https://github.com/eclipse/jetty.project/issues/1893), after upgrading we have tested the max duration implementation (https://github.com/eclipse/jetty.project/issues/5799) and we have encountered several issues by carrying out production load tests.

We have detected many `AsynchronousCloseException` after the connection was marked as removed due to expiration and after a certain time all our client requests fails with a `TimeoutException`. I don't know if this is two separated issues or the same root cause, but this is a blocking issue for us when all clients requests fails, the pool seems to be in a bad and is not able to recover.

**How to reproduce?**

I have reproduced the `AsynchronousCloseException` problem in the branch `11.0.x` by running the concurrent load test and completing the `HttpClientLoadTest.LoadTransportScenario.provideClientTransport()` with this snippet :

```java
                case H2:
                case H2C:
                {
                    ClientConnector clientConnector = new ClientConnector();
                    clientConnector.setSelectors(1);
                    clientConnector.setSslContextFactory(sslContextFactory);
                    HttpClientTransport clientTransport = new HttpClientTransportOverHTTP2(new HTTP2Client(clientConnector));
                    clientTransport.setConnectionPoolFactory(destination -> {
                        RoundRobinConnectionPool pool = new RoundRobinConnectionPool(destination, 4, destination);
                        pool.setMaxDuration(500);
                        return pool;
                    });
                    return clientTransport;
                }
```

The exception stacktrace :

```java
java.nio.channels.AsynchronousCloseException
	at org.eclipse.jetty.http2.client.http.HttpConnectionOverHTTP2.close(HttpConnectionOverHTTP2.java:200)
	at org.eclipse.jetty.client.HttpDestination.release(HttpDestination.java:427)
	at org.eclipse.jetty.http2.client.http.HttpChannelOverHTTP2.release(HttpChannelOverHTTP2.java:110)
	at org.eclipse.jetty.http2.client.http.HttpChannelOverHTTP2.exchangeTerminated(HttpChannelOverHTTP2.java:119)
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:473)
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:458)
	at org.eclipse.jetty.client.HttpReceiver.responseSuccess(HttpReceiver.java:421)
	at org.eclipse.jetty.http2.client.http.HttpReceiverOverHTTP2.access$100(HttpReceiverOverHTTP2.java:52)
	at org.eclipse.jetty.http2.client.http.HttpReceiverOverHTTP2$ContentNotifier.process(HttpReceiverOverHTTP2.java:298)
	at org.eclipse.jetty.http2.client.http.HttpReceiverOverHTTP2$ContentNotifier.offer(HttpReceiverOverHTTP2.java:259)
	at org.eclipse.jetty.http2.client.http.HttpReceiverOverHTTP2.notifyContent(HttpReceiverOverHTTP2.java:235)
	at org.eclipse.jetty.http2.client.http.HttpReceiverOverHTTP2.onData(HttpReceiverOverHTTP2.java:204)
	at org.eclipse.jetty.http2.client.http.HttpChannelOverHTTP2$Listener.onData(HttpChannelOverHTTP2.java:187)
	at org.eclipse.jetty.http2.api.Stream$Listener.onDataDemanded(Stream.java:256)
	at org.eclipse.jetty.http2.HTTP2Stream.notifyDataDemanded(HTTP2Stream.java:807)
	at org.eclipse.jetty.http2.HTTP2Stream.processData(HTTP2Stream.java:522)
	at org.eclipse.jetty.http2.HTTP2Stream.onData(HTTP2Stream.java:480)
	at org.eclipse.jetty.http2.HTTP2Stream.process(HTTP2Stream.java:368)
	at org.eclipse.jetty.http2.HTTP2Session.onData(HTTP2Session.java:262)
	at org.eclipse.jetty.http2.HTTP2Connection$ParserListener.onData(HTTP2Connection.java:407)
	at org.eclipse.jetty.http2.parser.BodyParser.notifyData(BodyParser.java:103)
	at org.eclipse.jetty.http2.parser.DataBodyParser.onData(DataBodyParser.java:145)
	at org.eclipse.jetty.http2.parser.DataBodyParser.onData(DataBodyParser.java:140)
	at org.eclipse.jetty.http2.parser.DataBodyParser.parse(DataBodyParser.java:106)
	at org.eclipse.jetty.http2.parser.Parser.parseBody(Parser.java:193)
	at org.eclipse.jetty.http2.parser.Parser.parse(Parser.java:122)
	at org.eclipse.jetty.http2.HTTP2Connection$HTTP2Producer.produce(HTTP2Connection.java:278)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produceTask(AdaptiveExecutionStrategy.java:446)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:239)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:190)
	at org.eclipse.jetty.http2.HTTP2Connection.produce(HTTP2Connection.java:208)
	at org.eclipse.jetty.http2.HTTP2Connection.onFillable(HTTP2Connection.java:155)
	at org.eclipse.jetty.http2.HTTP2Connection$FillableCallback.succeeded(HTTP2Connection.java:378)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:548)
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:397)
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:164)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
```

For reproducibility reason, the max duration is very low (500ms) in this test but in our production load tests the max duration parameter was setted to 60s and the problem was the same.

### PR DESCRIPTION
**Pull request by lorban**
The connection pool can close an expired multiplexed connection that one thread is done with but others are still using.

This fix ensures the good-to-be-closed advice given by the `Pool` object is correctly propagated when a connection expired such as the connection only gets closed when it has been marked as closed _and_ no thread uses it anymore.

Note: this PR also contains a behavioral change that makes the connection pool close expired connections also upon release instead of just during acquisition.

### CODE REVIEW

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 1]
```diff
@@ -85,14 +89,127 @@ public void disposeClient() throws Exception
...
+        // release req0 once req1 is done; req 1 should not have closed the response as req 0 is still running
+        assertTrue(reqClientDoneLatches[1].await(5, TimeUnit.SECONDS));
+        reqFinishingLatches[0].countDown();
+        assertTrue(reqClientDoneLatches[0].await(5, TimeUnit.SECONDS));
+
+        // Check that the pool created 2 and removed 2 connections;
+        // 2 were removed b/c waiting for req 2 means the 2nd connection
+        // expired and has to be removed and closed upon being returned to the pool.
+        assertThat(poolCreateCounter.get(), Matchers.is(2));
+        assertThat(poolRemoveCounter.get(), Matchers.is(2));
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java:192]: Please check also that the pool has 0 connections.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 2]
```diff
@@ -85,14 +89,127 @@ public void disposeClient() throws Exception
...
+            }
+        });
+
+        ClientConnector clientConnector = new ClientConnector();
+        HttpClientTransport transport = new HttpClientTransportOverHTTP2(new HTTP2Client(clientConnector));
+        transport.setConnectionPoolFactory(factory.factory);
+        client = new HttpClient(transport);
+        client.start();
+
+        CountDownLatch[] reqClientDoneLatches = new CountDownLatch[] {new CountDownLatch(1), new CountDownLatch(1), new CountDownLatch(1)};
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Rename to `reqClientSuccessLatches`.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 3]
```diff
@@ -321,7 +323,7 @@ protected Connection activate()
                     EntryHolder holder = (EntryHolder)((Attachable)connection).getAttachment();
                     if (holder.isExpired(maxDurationNanos))
                     {
-                        boolean canClose = remove(connection, true);
+                        boolean canClose = remove(connection);
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java:326]: Deprecate `remove(Connection, boolean)` and move its implementation in `remove(Connection)`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 1 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java:192]: Done.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 2 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Done.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 3 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java:326]: Done.

**Review by sbordet — APPROVED** [review]

**Pull request merged**